### PR TITLE
Pype 3.0 - command and install refactoring

### LINKED ISSUE #7: PYPE-663: separate venv for each deployed version
We need virtual environments for every pype installation, separate dev and prod branches for example.

Suggestions:

we can create hash for installation path and use that for PYPE_ENV path… So:

on linux pype env root: /opt/pype/ and we have pype dev version on /pipeline/pype/dev/pype-setup

we create hash from path, lets say it is fc85bc so complete PYPE_ENV for it will be at /opt/pype/fc85bc

[PYPE-663](https://pypeclub.atlassian.net/browse/PYPE-663)

### LINKED ISSUE #76: requirements are reporting inconsistent when git is present
<p>it would be great to be able to use git repositories in requirements. currently it will alway report inconsistent even though the latest is used</p>
<i>[PYPE-598] created by admin</i>

### LINKED ISSUE #176: 3.0: Move most code from pype-setup/pypeapp to pype
Move almost each file from pype-setup/pypeapp to pype/lib. Like anatomy, config, log, etc.

### LINKED ISSUE #211: Pype setup refactor
## Pype setup rework

For 3.0 release we change pype-setup so it acts as installer only, all other functionality should be moved to pype (see pypeclub/pype#205).

It should wrap all dependencies using something [cx_freeze](https://anthony-tuininga.github.io/cx_Freeze/).

### Local installation

How to tell the installer what to install? Compare *"local version"* with *"studio vesion"*
if they don't match update all repositories by copying from studio to local.

Installing Git and Python along with all other pype-setup files able to pack all dependent data files (json, ffmpeg, plugins_folder). We could append studio local sub-releases to `pype-setup/version.py`
everytime local changes should be pushed to artist, we add local version token to `pype-setup/config.ini`

#### Example:
```
3.0.0
3.0.0-client1 (version + local_subversion)
3.0.1-client1 (version + local_subversion)
3.0.2
3.1.1
3.1.1-client1 (version + local_subversion)
3.1.1-client2 (version + local_subversion)
```

Move `deploy.json` to `setup.py` and `config.ini`
Artist shouldn't need to update **pype-setup** itself often. Only with any major changes, in which case he get's new zip.

### PR DESCRIPTION
**Pull request by antirotor**
## Feature

This is first step towards new Pype 3 infrastructure. It dissolves **pypeclub/pype-setup** into **pypeclub/pype**. It provides basic build infrastructure to build/freeze Pype into standalone self-contained application. All required repositories are now added as git submodules into `repos` folder.

You can either clone it and init submodules in one step:
```
git clone -b 3.0/refactoring --recurse-submodules git@github.com:pypeclub/pype.git
```

Or if you already cloned pype just run:
```
git submodule update --init --recursive.
```

#### Windows

You will need [Python 3.7 and newer](https://www.python.org/downloads/) and [git](https://git-scm.com/downloads).

Clone repository:
```sh
git clone --recurse-submodules git@github.com:pypeclub/pype.git
```

Run PowerShell script `build.ps1`. It will create *venv*, install all
required dependencies and build Pype. After it is finished, you will find
Pype in `build` folder.

Pype is build using [CX_Freeze](https://cx-freeze.readthedocs.io/en/latest) to freeze itself and all dependencies.

Closes #211 
Closes #176
Closes #76 
Closes #7

### CODE REVIEW

**Comment by antirotor** [comment]
This doesn't work yet BTW. It should build OK under windows and print help, but `pype_commands` (former `PypeLauncher`) is not implemented yet.

**Comment by antirotor** [comment]
It is now possible to run **Tray**. Well - almost. What is missing is refactor of `pype.lib` **package** into `pype.lib` **module**.

File: pype/tools/tray/pype_tray.py [hunk 1]
```diff
@@ -221,7 +212,7 @@ def add_module(self, item, parent_menu):
                             "Module \"{}\" does not have attribute \"{}\"."
                             " Check your settings please."
                         ).format(import_path, key))
-
+            p = os.environ["AVALON_SCHEMA"]
```
[the remark this answers is not in the document]
Comment (iLLiCiTiT) [reply pype/tools/tray/pype_tray.py:215]: This should be removed? :D

File: pype/tools/tray/pype_tray.py [hunk 2]
```diff
@@ -1,5 +1,8 @@
 import os
 import sys
+
+s = dict(os.environ)
```
Comment (iLLiCiTiT) [inline pype/tools/tray/pype_tray.py:4]: This should be probably removed to

File: pype/tools/tray/__main__.py [hunk 3]
```diff
@@ -1,6 +1,7 @@
 import os
 import sys
-import pype_tray
+
+from . import pype_tray
```
Comment (iLLiCiTiT) [inline pype/tools/tray/__main__.py:4]: I think it is not allowed to do relative import in `__main__`

File: pype/tools/settings/__init__.py [hunk 4]
```diff
@@ -1,4 +1,4 @@
-from settings import style, MainWidget
+from .settings import style, MainWidget
```
Comment (iLLiCiTiT) [inline pype/tools/settings/__init__.py:1]: I think it is not allowed to do relative import here if launched from `__main__`.

File: pype/modules/ftrack/__init__.py [hunk 5]
```diff
@@ -1,7 +1,14 @@
+import os
+
 from . import ftrack_server
 from .ftrack_server import FtrackServer, check_ftrack_url
 from .lib import BaseHandler, BaseEvent, BaseAction, ServerAction
 
+from pype.api import system_settings
+
+os.environ["FTRACK_SERVER"] = (
```
Comment (iLLiCiTiT) [inline pype/modules/ftrack/__init__.py:9]: I would just add `# TODO set in ftrack module` to keep record that this is not place where this should happen

File: pype/tools/tray/__main__.py (hunk 3 above)
Comment (antirotor) [reply pype/tools/tray/__main__.py:4]: sure, but settings and tray are now launched like modules like `python -m pype.tools.tray`

File: pype/tools/settings/__init__.py (hunk 4 above)
Comment (antirotor) [reply pype/tools/settings/__init__.py:1]: sure, but settings and tray are now launched like modules like `python -m pype.tools.settings`

File: pype/tools/tray/__main__.py (hunk 3 above)
Comment (iLLiCiTiT) [reply pype/tools/tray/__main__.py:4]: In that case remove `__main__.py` files please.

File: pype/lib/applications.py [hunk 6]
```diff
@@ -213,7 +212,7 @@ class ApplicationAction(avalon.api.Action):
     @property
     def log(self):
         if self._log is None:
-            self._log = Logger().get_logger(self.__class__.__name__)
+            self._log = logging.getLogger(self.__class__.__name__)
```
Comment (iLLiCiTiT) [inline pype/lib/applications.py:215]: `ApplicationAction` class will be removed from here so it doesn't matter byt why changed from `Logger().get_logger` to `logging.getLogger`?

File: pype/tools/launcher/actions.py [hunk 7]
```diff
@@ -102,3 +102,7 @@ def register_environment_actions():
                         module, str(e)
                     )
                 )
+
+
+class ApplicationLaunchFailed(Exception):
```
Comment (iLLiCiTiT) [inline pype/tools/launcher/actions.py:107]: Why this exceptions is here? Should be in `pype.lib.applications`.

**Review by mkolar — APPROVED** [review]
Let's merge this and start cleaning up with all other code.

**Pull request merged**